tests/e2e-tests/specs/admin/shipping-settings.test.js

```diff
@@ -0,0 +1,348 @@
+ * @param {string} text The text value of the element we want to search for. ie. button's value, div's innertext.
+ */
+const waitForSelectorAndText = async (selector, text) => {
+    return await page.waitForFunction(
+        (cssSelector, innerTextContent) => !!Array.from(document.querySelectorAll(cssSelector)).find(el => el.textContent.trim() === innerTextContent.trim()),
+        {},
+        selector, text
+    );
+};
+
+describe( 'Saving shipping label settings', () => {
+	it( 'Can toggle shipping labels' , async () => {
+		await StoreOwnerFlow.login();
+		await StoreOwnerFlow.openSettings('shipping', 'woocommerce-services-settings');
+        await page.waitForSelector('.form-toggle__switch');
+        await expect(page).toClick('.form-toggle__switch');
+        await page.waitForSelector('.card.label-settings__labels-container', {
+            visible: false
+        });
+        await expect( page ).toClick( '.button.is-primary', { text: 'Save changes' } );
+        // Clicking save should persist the visible state of the section.
+        await page.waitForSelector('.card.label-settings__labels-container', {
+            visible: false
+        });
+
+        // Toggle it back to enable so the following tests can run.
+        await expect(page).toClick('.form-toggle__switch');
+        await saveAndWait();
+    });
+
+    it ('Should be able to select a different paper size', async () => {
+        // Save it as legal
+        await page.select('select.form-select', 'legal');
+        await saveAndWait();
+        await expect(page).toMatchElement('.notice.is-success .notice__text', { text: 'Your shipping settings have been saved.' });
+
+        let paperSize = await page.$('select.form-select');
+        let selectedOption = await (await paperSize.getProperty('value')).jsonValue();
+        expect(selectedOption).toBe('legal');
+
+        // Save it back to label
+        await page.select('select.form-select', 'label');
+        await saveAndWait();
+        await expect(page).toMatchElement('.notice.is-success .notice__text', { text: 'Your shipping settings have been saved.' });
+
+        paperSize = await page.$('select.form-select');
+        selectedOption = await (await paperSize.getProperty('value')).jsonValue();
+        expect(selectedOption).toBe('label');
+    });
+} );
+
+describe( 'Shipping label payment method', () => {
+    let response;
+
+    const accountSettingsRequestListener = (request) => {
+        if (request.url().match('wp-json/wc/v1/connect/account/settings')) {
+            request.respond({
+                status: 200,
+                contentType: 'application/json; charset=UTF-8',
+                body: JSON.stringify(response)
+            });
+        } else {
+            request.continue();
+        }
+    };
+
+    const mockAccountSettingAPI = async (mockResponse) => {
+        response = mockResponse;
+        page.on('request', accountSettingsRequestListener);
```

Comment:
I wonder what happens if an unhandled exception is generated during the test run. We should probably catch all these exceptions and remove the active listener, otherwise we're gonna get weird errors in the rest of the test run.

Reply:
Good question! The `removeListener` is inside `afterEach` of this `describe`. It should run and remove listeners even if the `it()` test cases throw exceptions.

Reply:
Just tested by adding a `throw new Error` in one test case, and counted 7 times `afterEach` calls, same count when I remove the throw error. I think we should be okay!

Reply:
awesome, thanks for checking this Harris! I wasn't aware the before/after hooks were called even after an unhandled error :+1: